FamiliyDoctor/FamiliyDoctor/src/main/java/com/familydoctor/services/DoctorServiceImpl.java

```diff
@@ -0,0 +1,141 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDate;
+import java.time.Period;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+@Service
+@Slf4j
+public class DoctorServiceImpl implements DoctorService {
+    private final DoctorRepository doctorRepository;
+    private final PatientRepository patientRepository;
+    private final ObjectMapper objectMapper;
+
+    @Autowired
+    public DoctorServiceImpl(DoctorRepository doctorRepository, PatientRepository patientRepository, ObjectMapper objectMapper) {
+        this.doctorRepository = doctorRepository;
+        this.patientRepository = patientRepository;
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public List<DoctorDTO> getAllDoctors() {
+        List<Doctor> doctors = doctorRepository.findAll();
+        return mapDoctorsToDTOs(doctors);
+    }
+
+    @Override
+    public DoctorDTO getDoctorById(Long id) {
+        Doctor doctor = doctorRepository.findById(id)
+                .orElseThrow(() -> new DoctorNotFoundException("Doctor not found: Doctor with id " + id, id));
+        return mapDoctorToDTO(doctor);
+    }
+
+    @Override
+    public DoctorDTO createDoctor(DoctorDTO doctorDTO) {
+        Doctor doctorToSave = mapDTOToDoctor(doctorDTO);
+        Doctor savedDoctor = doctorRepository.save(doctorToSave);
+        return mapDoctorToDTO(savedDoctor);
+    }
+
+    @Override
+    public DoctorDTO updateDoctor(Long id, DoctorDTO doctorDTO) {
+        Doctor existingDoctor = doctorRepository.findById(id)
+                .orElseThrow(() -> new DoctorNotFoundException("Doctor not found: Doctor with id " + id, id));
+        Doctor updatedDoctor = mapDTOToDoctor(doctorDTO);
+        updatedDoctor.setId(existingDoctor.getId());
+        Doctor savedDoctor = doctorRepository.save(updatedDoctor);
+        return mapDoctorToDTO(savedDoctor);
+    }
+
+    @Override
+    public void deleteDoctor(Long id) {
+        if (!doctorRepository.existsById(id)) {
+            throw new DoctorNotFoundException("Doctor not found: Doctor with id " + id, id);
+        }
+        doctorRepository.deleteById(id);
+    }
+    @Override
+    public PatientDTO addPatient(Long doctorId, PatientDTO patientDTO) {
+        Doctor doctor = doctorRepository.findById(doctorId)
+                .orElseThrow(() -> new DoctorNotFoundException("Doctor not found: Doctor with id " + doctorId, doctorId));
+
+        Patient patient = new Patient();
+        patient.setFirstName(patientDTO.getFirstName());
+        patient.setLastName(patientDTO.getLastName());
+        patient.setEmail(patientDTO.getEmail());
+        patient.setAddress(patientDTO.getAddress());
+
+        Random random = new Random();
+        long id = random.nextLong();
```

Comment:
Please review this entire logic. We never assign manually an id for an entity, we rely on the database to do that.